pkg/apiserver/controllers/tunnel_agent_reconciler.go

```diff
@@ -0,0 +1,316 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	controllerlog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	corev1alpha2 "github.com/apoxy-dev/apoxy/api/core/v1alpha2"
+	tunnet "github.com/apoxy-dev/apoxy/pkg/tunnel/net"
+	"github.com/apoxy-dev/apoxy/pkg/tunnel/vni"
+)
+
+const (
+	indexByTunnelRef = "spec.tunnelRef.name"
+)
+
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnelagents,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnelagents/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnelagents/finalizers,verbs=update
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnels,verbs=get;list;watch
+
+type TunnelAgentReconciler struct {
+	client    client.Client
+	agentIPAM tunnet.IPAM
+	vniPool   *vni.VNIPool
+}
+
+func NewTunnelAgentReconciler(c client.Client, agentIPAM tunnet.IPAM, vniPool *vni.VNIPool) *TunnelAgentReconciler {
+	return &TunnelAgentReconciler{
+		client:    c,
+		agentIPAM: agentIPAM,
+		vniPool:   vniPool,
+	}
+}
+
+func (r *TunnelAgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := controllerlog.FromContext(ctx, "name", req.Name)
+
+	var agent corev1alpha2.TunnelAgent
+	if err := r.client.Get(ctx, req.NamespacedName, &agent); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// handle deletion
+	if !agent.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(&agent, ApiServerFinalizer) {
+			if err := r.releaseResourcesIfPresent(log, &agent); err != nil {
+				log.Error(err, "failed to release resources; will retry")
+				return ctrl.Result{}, fmt.Errorf("failed to release resources: %w", err)
+			}
+
+			// Remove finalizer
+			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
nit: Probably don't need to retry considering `.releaseResourcesIfPresent` is idempotent - can just retry on next pass.